src/encoder_config.rs

```diff
@@ -0,0 +1,674 @@
+use core::mem::MaybeUninit;
+use core::ops::{Deref, DerefMut};
+
+use av_data::rational::Rational64;
+
+use crate::encoder::AV1Encoder;
+use crate::ffi::*;
+
+/// Encoder configuration structure
+///
+/// This structure contains the encoder settings that have common representations
+/// across all codecs. This doesn't imply that all codecs support all features,
+/// however.
+pub struct AV1EncoderConfig {
+    cfg: aom_codec_enc_cfg,
+}
+
+unsafe impl Send for AV1EncoderConfig {} // TODO: Make sure it cannot be abused
+
+impl AV1EncoderConfig {
+    /// Create a new default configuration
+    pub fn new() -> Result<AV1EncoderConfig, aom_codec_err_t::Type> {
+        let mut cfg = MaybeUninit::uninit();
+        let ret = unsafe { aom_codec_enc_config_default(aom_codec_av1_cx(), cfg.as_mut_ptr(), 0) };
+
+        match ret {
+            aom_codec_err_t::AOM_CODEC_OK => {
+                let cfg = unsafe { cfg.assume_init() };
+                Ok(AV1EncoderConfig { cfg })
+            }
+            _ => Err(ret),
+        }
+    }
+
+    /// Return a newly allocated `AV1Encoder` using the current configuration
+    pub fn get_encoder(&mut self) -> Result<AV1Encoder, aom_codec_err_t::Type> {
+        AV1Encoder::new(self)
+    }
+
+    /// Get a reference to the underlying config structure
+    pub fn cfg(&self) -> &aom_codec_enc_cfg {
+        &self.cfg
+    }
+
+    /// Get a mutable reference to the underlying config structure
+    pub fn cfg_mut(&mut self) -> &mut aom_codec_enc_cfg {
+        &mut self.cfg
+    }
+}
+
+/// # Generic settings (g)
+impl AV1EncoderConfig {
+    /// Algorithm specific "usage" value
+    ///
+    /// Algorithms may define multiple values for usage, which may convey the
+    /// intent of how the application intends to use the stream. If this value
+    /// is non-zero, consult the documentation for the codec to determine its
+    /// meaning.
```

Comment:
Since we support only av1, usage can be made an enum.

Reply:
Still working on this. Will address this last one sometime later today